### ISSUE TITLE
need ABI update for defaulted operator<=> implicitly declaring an operator==

### ISSUE THREAD

**Issue by zygoloid** [itanium-cxx-abi/cxx-abi#88]
Consider:

```
struct A {
  virtual operator<=>(const A&) const;
};
```

The vtable for `A` needs a slot for the virtual function `A::operator<=>` and a slot for the implicitly-declared virtual function `A::operator==`.

*Proposal:* allocate virtual table slots for virtual `operator==`s as if they were declared at the end of the class, in order of the declarations of the corresponding `operator<=>`s, after any virtual table slots inserted for implicitly-declared `operator=` functions.

**Comment by rjmccall** [comment]
How does this feature work?  What happens if there's an explicitly-declared or inherited `operator==`?

**Comment by zygoloid** [comment]
> Okay. And that declaration will be virtual if `operator<=>` is?

Yes.

> My instinct is to say that the operator== should come after the corresponding operator<=>, but that might be kindof painful.

We don't know whether any `operator==`s are declared until we get to the end of the class, and that's where they're notionally declared and checked (like with special members), so putting them at the end of the vtable seems likely to be easiest for implementations.

For this testcase:

```
struct X {
  virtual void x();
  friend auto operator<=>(X, X) = default;
};
struct Y {
  virtual ~Y();
  virtual A &operator=(const A&);
  friend auto operator<=>(Y, Y) = default;
};
struct A {
  virtual void f();
  virtual std::strong_ordering operator<=>(const A&) const& = default;
  virtual void g();
  virtual std::strong_ordering operator<=>(const A&) const&& = default;
  virtual void h();
};
```

The Clang 10 branch lays out the vtable for `A` as:

**X-in-A primary vtable**
  * offset to complete object
  * type info
  * `_ZN1X1xEv` `X::x`

**A extensions to primary vtable**
  * `_ZN1A1fEv` `A::f`
  * `_ZNKR1AssERKS_` `A::op<=>&`
  * `_ZN1A1gEv` `A::g`
  * `_ZNKO1AssERKS_` `A::op<=>&&`
  * `_ZN1A1hEv` `A::h`
  * `_ZN1AaSERKS_` `A::op=(const A&)`
  * `_ZNKR1AeqERKS_` `A::op==&`
  * `_ZNKO1AeqERKS_` `A::op==&&`
  * `_ZN1AD2Ev` `A::~A`
  * `_ZN1AD0Ev` `A::~A`

**Y-in-A vtable**
  * offset to complete object
  * type info
  * `_ZThn8_N1AD1Ev` thunk for `A::~A`
  * `_ZThn8_N1AD0Ev` thunk for `A::~A`
  * `_ZThn8_N1AaSERKS_` thunk for `A::op=(const A&)`

... so it is placing the `operator==`s between the `operator=` and the destructor. It seems reasonable to change that to put the `operator==`s at the end. (We should do that before clang 10 ships though!)

**Comment by rjmccall** [comment]
> the result of `<=>` and `==` should be expected to be consistent

Agreed, but... oh, I see, my bad assumption was that the defaulted definition of `operator==` would simply be to call `operator<=>`, whereas in fact it's derived fully independently.

> No attempt is ever made to infer an `operator==` from a user-defined `operator<=>` because there are cases where doing so is inefficient.

Ah, yes, that sounds typical.

> We don't know whether any `operator==`s are declared until we get to the end of the class, and that's where they're notionally declared and checked (like with special members), so putting them at the end of the vtable seems likely to be easiest for implementations.

Yes, that's probably true.

> so it is placing the `operator==`s between the `operator=` and the destructor.

See, the thing that worries me is that `Sema::AddImplicitlyDeclaredMembersToClass` very clearly adds declarations for the implicit equality operators *last*. :)  And maybe that's okay, and these implicit methods just need to be handled differently in v-table layout — except that while I see that there's special treatment of implicit destructors in `VTableBuilder.cpp`, there's no special treatment of any other implicit virtual functions.  The current Clang layout seems accidental.

**Comment by Quuxplusone** [comment]
@zygoloid: Your example code currently has a few typos — `A` for `Y`, and also `A` should inherit from both `X` and `Y`.

Is the virtual `operator=` in `Y` just a red herring? My understanding is that `operator=` behaves exactly like any old virtual function `foo`, and that there is no situation in C++17 nor in C++2a where the compiler ever needs to invent a vtable slot for a synthesized virtual member in the same way you're talking about it inventing a vtable slot for a synthesized virtual `operator==`. (My understanding is that in C++17, the only times `A` ever needs a vtable slot for a virtual member function `f` is when it explicitly declares `f`, or when one of its bases already has a vtable slot for `f`.) But if I'm wrong about that and `operator=` _is_ significant here, I'd like to learn!

(I note that polymorphic objects cannot meaningfully be compared anyway. This falls naturally out of not only how they're used, but the fact that C++ doesn't do covariant parameter types — once you have `virtual auto Base::operator<=>(const Base&) const`, there is no sensible way to override it. `auto Derived::operator<=>(const Derived&) const override` would just give an error, and `auto Derived::operator<=>(const Base&) const override` would be non-sensible, besides triggering `-Wambiguous-reversed-operator` when you tried to use it.)

**Comment by rjmccall** [comment]
@Quuxplusone The declaration of the assignment operator in `Y` is not a typo; Richard is declaring a non-standard assignment operator which will end up being overridden by the implicit copy assignment operator in `A`.  That this was possible had never before occurred to me, either. :)

The return type could, of course, also be declared `Y &` and the override would still be accepted as a covariant return.

You're right that Richard's example is missing the inheritance clause and a forward-declaration of `A`.

**Comment by Quuxplusone** [comment]
@rjmccall: Ah, I see how a forward declaration of `struct A` would fix it too. But in any case, all that matters to the C++17 ABI is that `Y` has a virtual function explicitly declared, and `A` overrides that function (so, the appropriate value gets filled into a vtable slot already allocated for that function in `vtable for Y in A`). It doesn't matter whether that function is named `operator=` or named `foo`, right?

**Comment by rjmccall** [comment]
At a language level, that is correct.  At an ABI level, overrides don't always simply change the function pointer in the existing v-table; they sometimes introduce new v-table entries.  Because of this, it is important to specify the order in which those entries are added.  It's sensible to follow source order where possible (and where guaranteed stable by the ODR, i.e. within a particular class definition), but since that isn't possible when we're talking about implicit declarations, we do need a special rule.  And (today at least — this will change if there are ever static facilities that can generate new members) implicit declarations can never be named `foo`.

**Comment by rjmccall** [comment]
@zygoloid, do you want to open a PR for this?
